File: frontend/test/features/calendar/calendar_grouping_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:smart_todo_app/features/calendar/domain/calendar_grouping.dart';
import 'package:smart_todo_app/features/task/domain/entities/task.dart';

Task _task(String id, {DateTime? due}) => Task(
      id: id,
      title: id,
      status: TaskStatus.todo,
      priority: TaskPriority.medium,
      dueDate: due,
      createdAt: DateTime.utc(2026, 7, 1),
      updatedAt: DateTime.utc(2026, 7, 1),
    );

void main() {
  test('buckets tasks by their local due day, ignoring the time', () {
    final map = groupTasksByDay([
      _task('a', due: DateTime(2026, 7, 18, 9)),
      _task('b', due: DateTime(2026, 7, 18, 23, 30)),
      _task('c', due: DateTime(2026, 7, 20, 12)),
    ]);

    expect(map[DateTime(2026, 7, 18)]!.map((t) => t.id), ['a', 'b']);
    expect(map[DateTime(2026, 7, 20)]!.single.id, 'c');
  });

  test('tasks without a due date are excluded', () {
    final map = groupTasksByDay([_task('a'), _task('b', due: DateTime(2026, 7, 18))]);

    expect(map.values.expand((e) => e).map((t) => t.id), ['b']);
  });

  test('tasksOn returns the tasks for a given day', () {
    final tasks = [
      _task('a', due: DateTime(2026, 7, 18, 9)),
      _task('c', due: DateTime(2026, 7, 20)),
    ];

    expect(tasksOn(tasks, DateTime(2026, 7, 18, 15)).single.id, 'a');
    expect(tasksOn(tasks, DateTime(2026, 7, 19)), isEmpty);
  });
}

File: webapp/src/features/calendar/calendar-grouping.test.ts
import { describe, expect, it } from 'vitest'
import type { Task } from '@/api/types'
import { groupTasksByDay, localDayKey, tasksOn } from '@/features/calendar/calendar-grouping'

function task(id: string, dueDate?: Date): Task {
  return {
    id,
    title: `Task ${id}`,
    status: 'TODO',
    priority: 'MEDIUM',
    tags: [],
    dueDate: dueDate?.toISOString(),
    createdAt: '2026-07-01T08:00:00Z',
    updatedAt: '2026-07-01T08:00:00Z',
  }
}

/** Data locale: è così che la vede l'utente, ed è ciò che conta sul calendario. */
function localDate(year: number, month: number, day: number, hour = 12) {
  return new Date(year, month - 1, day, hour)
}

describe('raggruppamento del calendario', () => {
  it('usa il giorno locale come chiave, non l istante', () => {
    expect(localDayKey(localDate(2026, 7, 5, 23))).toBe('2026-07-05')
    expect(localDayKey(localDate(2026, 12, 31, 0))).toBe('2026-12-31')
  })

  it('mette insieme i task dello stesso giorno, a qualunque ora', () => {
    const byDay = groupTasksByDay([
      task('a', localDate(2026, 7, 5, 8)),
      task('b', localDate(2026, 7, 5, 22)),
      task('c', localDate(2026, 7, 6, 9)),
    ])

    expect(byDay.get('2026-07-05')?.map((t) => t.id)).toEqual(['a', 'b'])
    expect(byDay.get('2026-07-06')?.map((t) => t.id)).toEqual(['c'])
  })

  it('tiene fuori i task senza scadenza: non hanno un posto sul calendario', () => {
    const byDay = groupTasksByDay([task('a'), task('b', localDate(2026, 7, 5))])

    expect([...byDay.values()].flat().map((t) => t.id)).toEqual(['b'])
  })

  it('funziona a cavallo di mese e di anno', () => {
    const byDay = groupTasksByDay([
      task('fine-mese', localDate(2026, 7, 31, 23)),
      task('inizio-mese', localDate(2026, 8, 1, 0)),
      task('capodanno', localDate(2027, 1, 1, 0)),
    ])

    expect(byDay.get('2026-07-31')?.map((t) => t.id)).toEqual(['fine-mese'])
    expect(byDay.get('2026-08-01')?.map((t) => t.id)).toEqual(['inizio-mese'])
    expect(byDay.get('2027-01-01')?.map((t) => t.id)).toEqual(['capodanno'])
  })

  it('tasksOn seleziona il giorno indicato, ignorando l ora', () => {
    const tasks = [
      task('a', localDate(2026, 7, 5, 8)),
      task('b', localDate(2026, 7, 5, 20)),
      task('c', localDate(2026, 7, 6)),
      task('senza-data'),
    ]

    expect(tasksOn(tasks, localDate(2026, 7, 5, 3)).map((t) => t.id)).toEqual(['a', 'b'])
    expect(tasksOn(tasks, localDate(2026, 7, 7)).map((t) => t.id)).toEqual([])
  })
})
